src/main/java/com/google/devtools/build/lib/skyframe/SequencedSkyframeExecutor.java

```diff
@@ -513,29 +515,40 @@ private void handleDiffsWithMissingDiffInformation(
         diffPackageRootsUnderWhichToCheck.add(pair.getFirst());
       }
 
-      EnumSet<FileType> fileTypesToCheck =
-          EnumSet.of(FileType.EXTERNAL_REPO, FileType.EXTERNAL_IN_MANAGED_DIRECTORY);
+      EnumSet<FileType> fileTypesToCheck = EnumSet.noneOf(FileType.class);
+      Iterable<SkyValueDirtinessChecker> dirtynessCheckers = ImmutableList.of();
+
+      if (!diffPackageRootsUnderWhichToCheck.isEmpty()) {
+        dirtynessCheckers = Iterables.concat(
+            dirtynessCheckers,
+            ImmutableList.of(new MissingDiffDirtinessChecker(diffPackageRootsUnderWhichToCheck)));
+      }
+      if (checkExternalFiles) {
+        dirtynessCheckers = Iterables.concat(dirtynessCheckers, customDirtinessCheckers);
+        fileTypesToCheck =
+            EnumSet.of(FileType.EXTERNAL_REPO, FileType.EXTERNAL_IN_MANAGED_DIRECTORY);
+        if (externalFilesKnowledge.tooManyNonOutputExternalFilesSeen) {
+          fileTypesToCheck.add(FileType.EXTERNAL);
+        }
```

Comment:
As I've noted elsewhere, general "external" files should not be mixed up with "external repository" files, they are unrelated.

Reply:
Oh! I thought I saw `externalFilesKnowledge.tooManyNonOutputExternalFilesSeen` set when I had too many files in external repositories. But if that isn't the case, then I'm happy to decouple.

Reply:
No, this should be only for EXTERNAL-type files, which are different from EXTERNAL_REPO-type files.